File: ssf_sdl.c
#include <SDL2/SDL.h>
/* #include <SDL2_ttf/SDL_ttf.h> */
/* #include <SDL2_image/SDL_image.h> */
/* #include <SDL2_mixer/SDL_mixer.h> */

#include "ssf.h"

void drawWireFrame(SDL_Renderer *r, const WireFrame *wf, const Point *p, int angle) {
  Point points[MAX_WIREFRAME_POINTS];
  double s = sin(angle * M_PI / 180);
  double c = cos(angle * M_PI / 180);
  for (int i=0; i<wf->pointCount; i++) {
    points[i].x = wf->points[i].x * c - wf->points[i].y * s + p->x;
    points[i].y = -(wf->points[i].x * s + wf->points[i].y * c) + p->y;
  }

  SDL_SetRenderDrawColor(r, wf->r, wf->g, wf->b, SDL_ALPHA_OPAQUE);
  for (int i=0; i<wf->lineCount; i++) {
    SDL_RenderDrawLine(r,
                       points[wf->lines[i].from].x, points[wf->lines[i].from].y,
                       points[wf->lines[i].to].x, points[wf->lines[i].to].y);
  }
}

void drawHexagon(SDL_Renderer *r, const Hexagon *h) {
  SDL_SetRenderDrawColor(r, 0, 255, 0, SDL_ALPHA_OPAQUE);
  for (int i=0; i<6; i++) {
    SDL_RenderDrawLine(r,
                       h->points[i].x, h->points[i].y,
                       h->points[(i+1)%6].x, h->points[(i+1)%6].y);
  }
}

void drawGame(SDL_Renderer *r, Game *g) {
  SDL_SetRenderDrawColor(r, 0, 0, 0, SDL_ALPHA_OPAQUE);
  SDL_RenderClear(r);

  drawHexagon(r, &g->bigHex);
  drawHexagon(r, &g->smallHex);
  if (g->ship.o.alive) {
    drawWireFrame(r, &shipWireFrame, &g->ship.o.position, g->ship.o.angle);
  }
  if (g->fortress.o.alive) {
    drawWireFrame(r, &fortressWireFrame, &g->fortress.o.position, g->fortress.o.angle);
  }

  for (int i=0; i<MAX_MISSILES; i++) {
    if (g->missiles[i].o.alive) {
      drawWireFrame(r, &missileWireFrame, &g->missiles[i].o.position, g->missiles[i].o.angle);
    }
  }
  for (int i=0; i<MAX_SHELLS; i++) {
    if (g->shells[i].o.alive) {
      drawWireFrame(r, &shellWireFrame, &g->shells[i].o.position, g->shells[i].o.angle);
    }
  }

}

void pumpAndWait(int until) {
  while (SDL_GetTicks() < until) {
    SDL_PumpEvents();
    SDL_Delay(1);
  }
}

KeySym getKeySym(const SDL_Event *e) {
  switch (e->key.keysym.sym) {
  case SDLK_w:
    return THRUST_KEY;
  case SDLK_a:
    return LEFT_KEY;
  case SDLK_d:
    return RIGHT_KEY;
  case SDLK_SPACE:
    return FIRE_KEY;
  default:
    return NO_KEY;
  }
}

void readEvents(Game *g) {
  SDL_Event e;
  KeySym key;
  while (SDL_PollEvent(&e) == 1) {
    switch (e.type) {
    case SDL_KEYDOWN:
      if (e.key.keysym.sym == SDLK_ESCAPE) exit(0);
      key = getKeySym(&e);
      if (key != NO_KEY) pressKey(g, key);
      break;
    case SDL_KEYUP:
      key = getKeySym(&e);
      if (key != NO_KEY) releaseKey(g, key);
      break;
    case SDL_QUIT:
      exit(0);
      break;
    }
  }
}

int main(int argc, char **argv) {
  SDL_Window *w;
  SDL_Renderer *r;
  Game *g;

  printf("game: %lu\n", sizeof(Game));
  printf("config: %lu\n", sizeof(Config));
  printf("ship: %lu\n", sizeof(Ship));
  printf("fortress: %lu\n", sizeof(Fortress));
  printf("missile: %lu\n", sizeof(Missile));
  printf("shell: %lu\n", sizeof(Shell));

  SDL_Init(SDL_INIT_VIDEO);

  w = SDL_CreateWindow("Space Fortress",
                       SDL_WINDOWPOS_CENTERED,
                       SDL_WINDOWPOS_CENTERED,
                       710, 600, 0);
  r = SDL_CreateRenderer(w, -1, SDL_RENDERER_ACCELERATED);


  int last;
  int now = SDL_GetTicks();
  printf("start\n");
  for (int i=0; i<2000; i++) {
    g = makeExplodeGame();
    while (!isGameOver(g)) stepOneTick(g, 33);
    freeGame(g);
  }
  printf("ellapse time: %d\n", SDL_GetTicks() - now);

  /* g = makeAutoTurnGame(); */
  g = makeExplodeGame();

  while (!isGameOver(g)) {
    last = now;
    now = SDL_GetTicks();
    readEvents(g);
    stepOneTick(g, now-last);
    drawGame(r, g);
    SDL_RenderPresent(r);
    pumpAndWait(now + 33);
  }

  free(g);

  return 0;
}

File: ssf_sdl.py
import math
import sdl2
import ssf

def drawWireFrame(r, wf, p, angle):
    s = math.sin(angle * math.pi / 180);
    c = math.cos(angle * math.pi / 180);
    points = []
    for i in range(wf.pointCount):
        np = ssf.Point()
        np.x = wf.points[i].x * c - wf.points[i].y * s + p.x
        np.y = -(wf.points[i].x * s + wf.points[i].y * c) + p.y
        points.append(np)

    sdl2.SDL_SetRenderDrawColor(r, wf.r, wf.g, wf.b, sdl2.SDL_ALPHA_OPAQUE)

    for i in range(wf.lineCount):
        sdl2.SDL_RenderDrawLine(r,
                                int(points[wf.lines[i]._from].x), int(points[wf.lines[i]._from].y),
                                int(points[wf.lines[i].to].x), int(points[wf.lines[i].to].y))

def drawHexagon(r, h):
    sdl2.SDL_SetRenderDrawColor(r, 0, 255, 0, sdl2.SDL_ALPHA_OPAQUE)
    for i in range(6):
        sdl2.SDL_RenderDrawLine(r,
                                int(h.points[i].x), int(h.points[i].y),
                                int(h.points[(i+1)%6].x), int(h.points[(i+1)%6].y))

def drawGame(r, g):
    sdl2.SDL_SetRenderDrawColor(r, 0, 0, 0, sdl2.SDL_ALPHA_OPAQUE)
    sdl2.SDL_RenderClear(r);

    drawHexagon(r, g.contents.bigHex);
    drawHexagon(r, g.contents.smallHex);

    if g.contents.ship.o.alive:
        drawWireFrame(r, ssf.shipWireFrame, g.contents.ship.o.position, g.contents.ship.o.angle)

    if g.contents.fortress.o.alive:
        drawWireFrame(r, ssf.fortressWireFrame, g.contents.fortress.o.position, g.contents.fortress.o.angle)

    for i in range(ssf.MAX_MISSILES):
        if g.contents.missiles[i].o.alive:
            drawWireFrame(r, ssf.missileWireFrame, g.contents.missiles[i].o.position, g.contents.missiles[i].o.angle)

    for i in range(ssf.MAX_SHELLS):
        if g.contents.shells[i].o.alive:
            drawWireFrame(r, ssf.shellWireFrame, g.contents.shells[i].o.position, g.contents.shells[i].o.angle)

def pumpAndWait(until):
    while sdl2.SDL_GetTicks() < until:
        sdl2.SDL_PumpEvents()
        sdl2.SDL_Delay(1)

def getKeySym(e):
    if e.key.keysym.sym == sdl2.SDLK_w:
        return ssf.THRUST_KEY
    elif e.key.keysym.sym == sdl2.SDLK_a:
        return ssf.LEFT_KEY
    elif e.key.keysym.sym == sdl2.SDLK_d:
        return ssf.RIGHT_KEY
    elif e.key.keysym.sym == sdl2.SDLK_SPACE:
        return ssf.FIRE_KEY
    else:
        return ssf.NO_KEY

def readEvents(g):
    e = sdl2.SDL_Event()
    while sdl2.SDL_PollEvent(e) == 1:
        if e.type == sdl2.SDL_KEYDOWN:
            k = getKeySym(e)
            if k != ssf.NO_KEY:
                ssf.pressKey(g, k)
        elif e.type == sdl2.SDL_KEYUP:
            k = getKeySym(e)
            if k != ssf.NO_KEY:
                ssf.releaseKey(g, k)
        elif e.type == sdl2.SDL_QUIT:
            exit(0)

if __name__ == '__main__':
    sdl2.SDL_Init(sdl2.SDL_INIT_VIDEO)

    w = sdl2.SDL_CreateWindow("Space Fortress",
                              sdl2.SDL_WINDOWPOS_CENTERED,
                              sdl2.SDL_WINDOWPOS_CENTERED,
                              710, 600, 0)
    r = sdl2.SDL_CreateRenderer(w, -1, sdl2.SDL_RENDERER_ACCELERATED)

    now = sdl2.SDL_GetTicks()
    g = ssf.makeExplodeGame()
    while not ssf.isGameOver(g):
        last = now
        now = sdl2.SDL_GetTicks()
        readEvents(g)
        ssf.stepOneTick(g, now-last)
        drawGame(r, g)
        sdl2.SDL_RenderPresent(r)
        pumpAndWait(now + 33)

    print ssf.dumpSexpGameState(g)
